tests/zfs-tests/tests/functional/removal/removal_resume_export.ksh

```diff
@@ -105,20 +106,26 @@ log_must zfs set compression=off $TESTPOOL/$TESTFS
 log_must dd if=/dev/urandom of=$TESTDIR/$TESTFILE0 bs=64M count=32
 
 #
-# Add second device where all the data will be evacuated.
+# Unmount all filesystems now to minimize the time took by 'zpool export' later:
+# this helps avoid race conditions
 #
-log_must zpool add -f $TESTPOOL $NOTREMOVEDISK
+log_must zfs unmount -a
 
 #
-# Start removal.
+# Add second device where all the data will be evacuated.
 #
-log_must zpool remove $TESTPOOL $REMOVEDISK
+log_must zpool add -f $TESTPOOL $NOTREMOVEDISK
 
 #
 # Inject an error so export fails after having just suspended
 # the removal thread. [spa_inject_ref gets incremented]
 #
-log_must zinject -d $REMOVEDISK -D 10:1 $TESTPOOL
+log_must zinject -d $NOTREMOVEDISK -D 10:1 $TESTPOOL
+
+#
+# Start removal asynchronously
+#
```

Comment:
Why are we doing this in the background?  Seems like that will introduce a race condition in that `ensure_thread_running` may fail because we haven't yet started the removal.

Reply:
The `ensure_thread_running` helper function was implemented differently on Linux since it depended on `mdb`.  The updated implementation polls briefly waiting for the expected process before failing, so there shouldn't be an issue running this in the background.

This issue here was that we were losing the race in other direction.  The removal thread could already have started and have finished before we had a chance to check for it.  I'd forgotten all about `zfs_remove_max_bytes_pause` since it's only used in one other test, that does sound like a more generic way to test this.

Reply:
Thanks, i did not know about `zfs_remove_max_bytes_pause`, but now i don't understand why we need `ensure_thread_running` in this test when we can do the following:

```shell
function callback
{
        log_mustnot zpool export $TESTPOOL
        return 0
}
...
log_must zpool add -f $TESTPOOL $NOTREMOVEDISK
log_must zinject -d $REMOVEDISK -D 10:1 $TESTPOOL
log_must attempt_during_removal $TESTPOOL $REMOVEDISK callback
```

It seems using `attempt_during_removal()` will:

1. start the removal process
2. verify it's running
3. run the callback function (`log_mustnot zpool export $TESTPOOL`)
4. verify the removal is still running
5. resume the removal
6. verify the removal complete successfully

which is, unless i am missing something, all we need to run this test.

Reply:
@loli10K That sounds right, although IIRC `attempt_during_removal` is like a retry loop that doesn't use `zfs_remove_max_bytes_pause` so it's also a bit racy.  But it may be ironed out well enough to use as you're proposing.  But:
"4. verify the removal is still running" would presumably do this by calling `ensure_thread_running`
what do you mean by "5. resume the removal" since you are never pausing it?

@jwk404 any thoughts?

Reply:
`attempt_during_removal()` seems to use `zfs_remove_max_bytes_pause`, see https://github.com/zfsonlinux/zfs/blob/master/tests/zfs-tests/tests/functional/removal/removal.kshlib#L58

```shell
function attempt_during_removal # pool disk callback [args]
{
        typeset pool=$1
        typeset disk=$2
        typeset callback=$3

        shift 3
        set_tunable64 zfs_remove_max_bytes_pause 0

        log_must zpool remove $pool $disk # 1. start the removal process

        #
        # We want to make sure that the removal started
        # before issuing the callback.
        #
        sync
        log_must is_pool_removing $pool # 2. verify it's running

        log_must $callback "$@" # 3. run the callback function

        #
        # Ensure that we still haven't finished the removal
        # as expected.
        #
        log_must is_pool_removing $pool # 4. verify the removal is still running

        set_tunable64 zfs_remove_max_bytes_pause 18446744073709551615 # 5. resume the removal

        log_must wait_for_removal $pool # 6. verify the removal complete successfully
        log_mustnot vdevs_in_pool $pool $disk
        return 0
}
```

I think we can just remove `ensure_thread_running()` from this test and use the existing `attempt_during_removal()` from "removal.kshlib".

Reply:
Ah, wonderful.